enh-ruby-mode.el

```diff
@@ -1439,6 +1436,42 @@ With ARG, do it that many times."
       (setq erm-full-parse-p t)
       (error "%s" (substring response 1))))))
 
+(defun erm--end-p ()
+  "Is point directly after a block closing \"end\""
+  (let ((end-pos (- (point) 3)))
+    (and (>= end-pos (point-min)) (string= "end" (buffer-substring end-pos (point))) (eq (get-text-property end-pos 'indent) 'e))))
+
+
+(defun erm--advise-show-paren-data-function (orig &rest args)
+  ;; First check if we are on opening ('b or 'd). We only care about
+  ;; the word openers "if", "do" etc (normal show-paren handles "{")
+  (if (and (memq (get-text-property (point) 'indent) '(b d))
+           (looking-at "\\w"))
+      (save-excursion
+        (let ((opener-beg (point))
+              (opener-end (save-excursion (forward-word) (point)))
+              (closer-end (progn (enh-ruby-forward-sexp 1) (point))))
+          (list
+           opener-beg
+           opener-end
+           (save-excursion (skip-syntax-backward "(w") (point))
+           closer-end
+           (not (erm--end-p)))))
+    ;; Now check if we are at a closer ("end")
+    (if (erm--end-p)
+        (let ((end-pos (point)))
+          (save-excursion
+            (enh-ruby-backward-sexp 1)
+            (list
+             (- end-pos 3)
+             end-pos
+             (point)
+             (save-excursion (skip-syntax-forward "(w") (point))
+             (or (not (looking-at "\\w"))
+                 (not (memq (get-text-property (point) 'indent) '(b d)))))))
+      ;; Otherwise, delegate to normal show-paren-data-function
+      (apply orig args))))
```

Comment:
Is this advising for some reason? What if the user has changed it globally but you're expecting `show-paren--default` behavior? I think I'd rather this just be a normal function that calls through to `show-paren--default` as needed. I'd also like it to have tests. Since it just returns data, it should also be pretty easy to do.

Reply:
I did exactly what ruby-mode does to hook into show-paren-mode. If it was not advice, when would my function be invoked? show-paren-mode sets up an idle timer to check if point is near parens, so by using advice we hook into the same timer, but just add erm specific logic around the built in paren and curly logic.

Reply:
Your function could be invoked by being pointed to by `show-paren-data-function`, which is can be set by enh-ruby-mode to be a buffer-local variable.

Reply:
I'll commit this separately as this PR is already merged, but this is what I mean:

```diff
diff --git a/enh-ruby-mode.el b/enh-ruby-mode.el
index 6daa57a..a309ddb 100644
--- a/enh-ruby-mode.el
+++ b/enh-ruby-mode.el
@@ -360,6 +360,7 @@ Warning: does not play well with electric-indent-mode.
   (setq-local require-final-newline        mode-require-final-newline)
   (setq-local beginning-of-defun-function  'enh-ruby-beginning-of-defun)
   (setq-local end-of-defun-function        'enh-ruby-end-of-defun)
+  (setq-local show-paren-data-function     #'erm-show-paren-data-function)
 
   (setq-local paragraph-start
               (concat "$\\|" page-delimiter))
@@ -377,9 +378,6 @@ Warning: does not play well with electric-indent-mode.
   (add-hook 'change-major-mode-hook 'erm-major-mode-changed     nil t)
   (add-hook 'kill-buffer-hook       'erm-buffer-killed          nil t)
 
-  (add-function :around (local 'show-paren-data-function)
-                #'erm--advise-show-paren-data-function)
-
   (abbrev-mode)
   (erm-reset-buffer))
 
@@ -1443,7 +1441,7 @@ With ARG, do it that many times."
          (string= "end" (buffer-substring end-pos (point)))
          (eq (get-text-property end-pos 'indent) 'e))))
 
-(defun erm--advise-show-paren-data-function (orig &rest args)
+(defun erm-show-paren-data-function ()
   ;; First check if we are on opening ('b or 'd). We only care about
   ;; the word openers "if", "do" etc (normal show-paren handles "{")
   (if (and (memq (get-text-property (point) 'indent) '(b d))
@@ -1470,8 +1468,7 @@ With ARG, do it that many times."
              (save-excursion (skip-syntax-forward "(w") (point))
              (or (not (looking-at "\\w"))
                  (not (memq (get-text-property (point) 'indent) '(b d)))))))
-      ;; Otherwise, delegate to normal show-paren-data-function
-      (apply orig args))))
+      (show-paren--default))))
 
 (erm-reset)
```

Reply:
Ok, I see. I didn't realize show-paren-data-function was a variable meant to be set.